### PR TITLE
fix colorlog compatibility problem

### LINKED ISSUE #4403: #4376 breaks compatibility with old `colorlog`
### Expected behavior

`import optuna` should not fail.

### Environment

- Optuna version:
- Python version:
- OS:
- (Optional) Other libraries and their versions: **colorlog v4.8**

### Error messages, stack traces, or logs

```shell
File ~/optuna/optuna/logging.py:36, in create_default_formatter()
     31 def create_default_formatter() -> colorlog.ColoredFormatter:
     32     """Create a default formatter of log messages.
     33 
     34     This function is not supposed to be directly accessed by library users.
     35     """
---> 36     return colorlog.ColoredFormatter(
     37         "%(log_color)s[%(levelname)1.1s %(asctime)s]%(reset)s %(message)s",
     38         no_color=False if _color_supported() else True,
     39     )

TypeError: ColoredFormatter.__init__() got an unexpected keyword argument 'no_color'
```

### Steps to reproduce

1.
```
$ pip install colorlog==4.8.0
```
2.
```
import optuna
```

### Additional context (optional)
Updating to the latest `colorlog` fixes this.

### PR DESCRIPTION
**Pull request by gen740**
## Motivation

Fix #4403.

## Description of the changes

Use logging.Formatter if `_color_supported` returns True.

### CODE REVIEW

File: optuna/logging.py [hunk 1]
```diff
@@ -28,15 +30,16 @@
 _default_handler: Optional[logging.Handler] = None
 
 
-def create_default_formatter() -> colorlog.ColoredFormatter:
+def create_default_formatter() -> Union[Formatter, colorlog.ColoredFormatter]:
```
Comment (toshihikoyanase) [inline optuna/logging.py:33]: `colorlog.ColoredFormatter` extends `logging.Formatter`, and I guess the return type can be `Formatter`.

```suggestion
def create_default_formatter() -> Formatter:
```

And, the returned value is pass to `logging.Handler.setFormatter`, and it expects `logging.Formatter`. So, I don't think it does not depend on the `colorlog` version.

https://github.com/optuna/optuna/blob/ea55b34f9205d3e52fd4d825d4bf89c9daaf6696/optuna/logging.py#L70

File: optuna/logging.py [hunk 2]
```diff
@@ -28,15 +30,16 @@
...
     """
-    return colorlog.ColoredFormatter(
-        "%(log_color)s[%(levelname)1.1s %(asctime)s]%(reset)s %(message)s",
-        no_color=False if _color_supported() else True,
-    )
+    if _color_supported():
+        return colorlog.ColoredFormatter(
+            "%(log_color)s[%(levelname)1.1s %(asctime)s]%(reset)s %(message)s",
+        )
+    return Formatter("[%(levelname)1.1s %(asctime)s] %(message)s")
```
Comment (toshihikoyanase) [inline optuna/logging.py:42]: I'm not sure, but do you think we can simplify the format strings by extracting the common parts?

```suggestion
    header = "[%(levelname)1.1s %(asctime)s]"
    message = "%(message)s"
    if _color_supported():
        return colorlog.ColoredFormatter(
            f"%(log_color)s{header}%(reset)s {message}",
        )
    return Formatter(f"{header} {message}")
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your PR. I have two comments.

**Comment by toshihikoyanase** [comment]
@contramundum53 Could you review this PR as an author of https://github.com/optuna/optuna/issues/4403?

**Comment by contramundum53** [comment]
I confirmed that it works with colorlog 4.8. Thanks for the PR!

**Review by contramundum53 — APPROVED** [review]

**Comment by contramundum53** [comment]
@keisuke-umezawa Could you review this PR?

File: optuna/logging.py [hunk 3]
```diff
@@ -28,15 +28,18 @@
...
-        "%(log_color)s[%(levelname)1.1s %(asctime)s]%(reset)s %(message)s",
-        no_color=False if _color_supported() else True,
-    )
+    header = "[%(levelname)1.1s %(asctime)s]"
+    message = "%(message)s"
+    if _color_supported():
+        return colorlog.ColoredFormatter(
+            f"%(log_color)s{header}%(reset)s {message}",
+        )
+    return logging.Formatter("[%(levelname)1.1s %(asctime)s] %(message)s")
```
Comment (toshihikoyanase) [inline optuna/logging.py:42]: Thank you for your update. I think we can use `header` and `message` here.
```suggestion
    return logging.Formatter(f"{header} {message}")
```

Comment (gen740) [reply optuna/logging.py:42]: Oops, I forgot to change it. Thank you!

**Review by keisuke-umezawa — APPROVED** [review]
LGTM! I confirmed it works on my local env for both situations. Thank you for your quick fix.

**Pull request merged**